File: app/src/main/java/com/example/uscatterbrain/network/BlockHeaderPacket.java
package com.example.uscatterbrain.network;

import android.util.Log;

import com.example.uscatterbrain.ScatterProto;
import com.google.protobuf.ByteString;
import com.google.protobuf.GeneratedMessageLite;
import com.goterl.lazycode.lazysodium.interfaces.Sign;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;

/**
 * Wrapper class for protocol buffer blockdata message
 */
public class BlockHeaderPacket implements ScatterSerializable {
    private ScatterProto.BlockData blockdata;
    private List<ByteString> mHashList;
    private ByteString mFromFingerprint;
    private ByteString mToFingerprint;
    private byte[] mSignature;
    private byte[] mApplication;
    private int mSessionID;
    private boolean mToDisk;
    private int mBlocksize;

    private BlockHeaderPacket(Builder builder) {
        this.mHashList = builder.getHashlist();
        if (builder.getSig() == null) {
            this.mSignature = new byte[Sign.ED25519_BYTES];
        } else {
            this.mSignature = builder.getSig().toByteArray();
        }
        this.mToFingerprint = builder.getmToFingerprint();
        this.mFromFingerprint = builder.getmFromFingerprint();
        this.mApplication = builder.getApplication();
        this.mSessionID = builder.getSessionid();
        this.mToDisk = builder.getToDisk();
        this.mBlocksize = builder.getBlockSize();
        Log.e("debug", "adding all nexthashes " + mHashList.size());
        this.blockdata = ScatterProto.BlockData.newBuilder()
                .setApplicationBytes(ByteString.copyFrom(this.mApplication))
                .setFromFingerprint(this.mFromFingerprint)
                .setToFingerprint(this.mToFingerprint)
                .setTodisk(this.mToDisk)
                .addAllNexthashes(this.mHashList)
                .setSessionid(this.mSessionID)
                .setSig(ByteString.copyFrom(this.mSignature))
                .setBlocksize(mBlocksize)
                .build();
    }

    private ByteString sumBytes() {
        ByteString messagebytes = ByteString.EMPTY;
        messagebytes = messagebytes.concat(this.mFromFingerprint);
        messagebytes = messagebytes.concat(this.mToFingerprint);
        messagebytes = messagebytes.concat(ByteString.copyFrom(this.mApplication));
        byte td = 0;
        if (this.mToDisk)
            td = 1;
        ByteString toDiskBytes = ByteString.copyFrom(ByteBuffer.allocate(1).order(ByteOrder.BIG_ENDIAN).put(td).array());
        messagebytes = messagebytes.concat(toDiskBytes);

        for (ByteString hash : this.mHashList) {
            messagebytes = messagebytes.concat(hash);
        }
        return messagebytes;
    }

    /**
     * Verifyed 25519 boolean.
     *
     * @param pubkey the pubkey
     * @return the boolean
     */
    public boolean verifyed25519(byte[] pubkey) {
        if (pubkey.length != Sign.PUBLICKEYBYTES)
            return false;

        ByteString messagebytes = sumBytes();
        return LibsodiumInterface.getSodium().crypto_sign_verify_detached(this.blockdata.getSig().toByteArray(),
                messagebytes.toByteArray(),
                messagebytes.size(),
                pubkey) == 0;
    }

    /**
     * Sign ed 25519 boolean.
     *
     * @param secretkey the secretkey
     * @return the boolean
     */
    public boolean signEd25519(byte[] secretkey) {
        if (secretkey.length != Sign.SECRETKEYBYTES)
            return false;

        ByteString messagebytes = sumBytes();

        this.mSignature = new byte[Sign.ED25519_BYTES];
        Pointer p = new PointerByReference(Pointer.NULL).getPointer();
        if (LibsodiumInterface.getSodium().crypto_sign_detached(this.mSignature,
                p, messagebytes.toByteArray(), messagebytes.size(), secretkey) == 0) {
            this.blockdata = ScatterProto.BlockData.newBuilder()
                    .setApplicationBytes(ByteString.copyFrom(this.mApplication))
                    .setFromFingerprint(this.mFromFingerprint)
                    .setToFingerprint(this.mToFingerprint)
                    .setTodisk(this.mToDisk)
                    .addAllNexthashes(this.mHashList)
                    .setSessionid(this.mSessionID)
                    .setBlocksize(mBlocksize)
                    .setSig(ByteString.copyFrom(this.mSignature))
                    .build();
            return true;
        } else {
            return false;
        }
    }

    private BlockHeaderPacket(InputStream in) throws IOException {
        this.blockdata = ScatterProto.BlockData.parseDelimitedFrom(in);
        this.mApplication = blockdata.getApplicationBytes().toByteArray();
        Log.e("debug ", "header nexthashes count" + blockdata.getNexthashesList().size());
        Log.e("debug", "header nexthashes raw count " + blockdata.getNexthashesCount());
        this.mHashList = blockdata.getNexthashesList();
        this.mFromFingerprint = blockdata.getFromFingerprint();
        this.mToFingerprint = blockdata.getToFingerprint();
        this.mSignature = blockdata.getSig().toByteArray();
        this.mToDisk = blockdata.getTodisk();
        this.mSessionID = blockdata.getSessionid();
        this.mBlocksize = blockdata.getBlocksize();
    }

    /**
     * Parse from blockheader packet.
     *
     * @param is the is
     * @return the block header packet
     */
    public static Single<BlockHeaderPacket> parseFrom(InputStream is) {
        return Single.fromCallable(() -> new BlockHeaderPacket(is));
    }

    @Override
    public byte[] getBytes() {
        ByteArrayOutputStream os = new ByteArrayOutputStream();
        try {
            this.blockdata.writeDelimitedTo(os);
            return os.toByteArray();
        } catch (IOException e) {
            return null;
        }
    }

    @Override
    public ByteString getByteString() {
        return ByteString.copyFrom(getBytes());
    }

    @Override
    public Completable writeToStream(OutputStream os) {
        return Completable.fromAction(() -> blockdata.writeDelimitedTo(os));
    }

    @Override
    public GeneratedMessageLite getMessage() {
        return blockdata;
    }

    /**
     * Gets blockdata.
     *
     * @return the blockdata
     */
    public ScatterProto.BlockData getBlockdata() {
        return this.blockdata;
    }

    /**
     * Gets the blocksize
     * @return int blocksize
     */
    public int getBlockSize() {
        return this.blockdata.getBlocksize();
    }

    /**
     * Gets hash.
     *
     * @param seqnum the seqnum
     * @return the hash
     */
    public ByteString getHash(int seqnum) {
        return this.blockdata.getNexthashes(seqnum);
    }

    /**
     * Gets sig.
     *
     * @return the sig
     */
    public ByteString getSig() {
        return ByteString.copyFrom(this.mSignature);
    }

    /**
     * Get application byte [ ].
     *
     * @return the byte [ ]
     */
    public byte[] getApplication() {
        return this.mApplication;
    }

    /**
     * Gets hash list.
     *
     * @return the hash list
     */
    public List<ByteString> getHashList() {
        return mHashList;
    }

    /**
     * Gets from fingerprint.
     *
     * @return the from fingerprint
     */
    public ByteString getFromFingerprint() {
        return mFromFingerprint;
    }

    /**
     * Gets to fingerprint.
     *
     * @return the to fingerprint
     */
    public ByteString getToFingerprint() {
        return mToFingerprint;
    }

    /**
     * Get signature byte [ ].
     *
     * @return the byte [ ]
     */
    public byte[] getSignature() {
        return mSignature;
    }

    /**
     * Gets session id.
     *
     * @return the session id
     */
    public int getSessionID() {
        return mSessionID;
    }

    /**
     * Gets to disk.
     *
     * @return the to disk
     */
    public boolean getToDisk() {
        return mToDisk;
    }

    /**
     * Sets to disk.
     *
     * @param t the t
     */
    public void setToDisk(boolean t) { this.mToDisk = t; }

    /**
     * New builder builder.
     *
     * @return the builder
     */
    public static Builder newBuilder() {
        return new Builder();
    }

    /**
     * The type Builder.
     */
    public static class Builder {
        private boolean todisk;
        private byte[] application;
        private int sessionid;
        public int mBlockSize;
        private ByteString mToFingerprint;
        private ByteString mFromFingerprint;
        private List<ByteString> hashlist;
        private ByteString mSig;

        /**
         * Instantiates a new Builder.
         */
        public Builder() {
            todisk = false;
            sessionid = -1;
            mBlockSize = -1;
        }

        /**
         * Sets the fingerprint for the recipient.
         *
         * @param toFingerprint the to fingerprint
         * @return builder
         */
        public Builder setToFingerprint(ByteString toFingerprint) {
            this.mToFingerprint = toFingerprint;
            return this;
        }

        /**
         * Sets from fingerprint.
         *
         * @param fromFingerprint sets the fingerprint for the sender
         * @return builder
         */
        public Builder setFromFingerprint(ByteString fromFingerprint) {
            this.mFromFingerprint = fromFingerprint;
            return this;
        }

        /**
         * Sets application.
         *
         * @param application bytes for UTF encoded scatterbrain application string
         * @return builder
         */
        public Builder setApplication(byte[] application) {
            this.application = application;
            return this;
        }

        /**
         * Sets to disk.
         *
         * @param toDisk whether to write this file to disk or attempt to store it in the database
         * @return builder
         */
        public Builder setToDisk(boolean toDisk) {
            this.todisk = toDisk;
            return this;
        }

        /**
         * Sets session id.
         *
         * @param sessionID the session id (used for upgrading between protocols)
         * @return builder
         */
        public Builder setSessionID(int sessionID) {
            this.sessionid = sessionID;
            return this;
        }

        /**
         * Sets hashes.
         *
         * @param hashes list of hashes of following blocksequence packets.
         * @return builder
         */
        public Builder setHashes(List<ByteString> hashes) {
            this.hashlist = hashes;
            return this;
        }

        /**
         * Sets blocksize
         * @param blockSize
         * @return builder
         */
        public Builder setBlockSize(int blockSize) {
            this.mBlockSize = blockSize;
            return this;
        }

        public Builder setSig(ByteString sig) {
            this.mSig = sig;
            return this;
        }

        public ByteString getSig() {
            return mSig;
        }

        /**
         * Build block header packet.
         *
         * @return the block header packet
         */
        public BlockHeaderPacket build() {
            if (hashlist == null)
                return null;

            // fingerprints and application are required
            if (mFromFingerprint == null || mToFingerprint == null || application == null) {
                return null;
            }

            if (mBlockSize <= 0) {
                return null;
            }

            return new BlockHeaderPacket(this);
        }

        /**
         * Get application name (in UTF-8).
         *
         * @return the byte [ ]
         */
        public byte[] getApplication() {
            return application;
        }

        /**
         * Gets sessionid.
         *
         * @return the sessionid
         */
        public int getSessionid() {
            return sessionid;
        }

        /**
         * Gets hashlist.
         *
         * @return the hashlist
         */
        public List<ByteString> getHashlist() {
            return hashlist;
        }

        /**
         * Gets to fingerprint.
         *
         * @return the to fingerprint
         */
        public ByteString getmToFingerprint() {
            return mToFingerprint;
        }

        /**
         * Gets from fingerprint.
         *
         * @return the from fingerprint
         */
        public ByteString getmFromFingerprint() {
            return mFromFingerprint;
        }

        /**
         * Gets to disk.
         *
         * @return the to disk
         */
        public boolean getToDisk() {
            return todisk;
        }

        /**
         * Gets blocksize
         * @return blocksize
         */
        public int getBlockSize() { return mBlockSize; }
    }
}

File: app/src/main/java/net/ballmerlabs/uscatterbrain/network/BlockHeaderPacket.kt
package net.ballmerlabs.uscatterbrain.network

import android.util.Log
import com.google.protobuf.ByteString
import net.ballmerlabs.uscatterbrain.ScatterProto.BlockData
import net.ballmerlabs.uscatterbrain.db.getDefaultFileName
import net.ballmerlabs.uscatterbrain.db.isValidFilename
import net.ballmerlabs.uscatterbrain.db.sanitizeFilename
import java.util.*
import kotlin.collections.ArrayList

/**
 * Wrapper class for protocol buffer blockdata message
 */
class BlockHeaderPacket(blockdata: BlockData) : ScatterSerializable<BlockData>(blockdata), Verifiable  {
    /**
     * Gets hash list.
     *
     * @return the hash list
     */
    val hashList: List<ByteArray>
    get() = packet.nexthashesList.map { v -> v.toByteArray() }

    override val hashes
        get() = hashList.toTypedArray()


    /**
     * Gets from fingerprint.
     *
     * @return the from fingerprint
     */
    override val fromFingerprint: List<UUID>
        get() = packet.fromFingerprintList.map { f -> protoUUIDtoUUID(f) }

    override val sendDate: Long
        get() = packet.sendDate

    /**
     * Gets to fingerprint.
     *
     * @return the to fingerprint
     */
    override val toFingerprint: List<UUID>
        get() = packet.toFingerprintList.map { f -> protoUUIDtoUUID(f)}

    override val extension: String
        get() = sanitizeFilename(packet.extension)

    /**
     * Get signature byte [ ].
     *
     * @return the byte [ ]
     */
    override val signature: ByteArray?
        get() = if (fromFingerprint == null) null else packet.sig.toByteArray()

    /**
     * Get application byte [ ].
     *
     * @return the byte [ ]
     */
    override val application: String
        get() = packet.application

    val isValidFilename: Boolean
    get() = isValidFilename(packet.filename)

    /**
     * Gets session id.
     *
     * @return the session id
     */
    val sessionID: Int
        get() = packet.sessionid

    override val toDisk: Boolean
        get() = packet.todisk

    val isEndOfStream: Boolean
        get() = packet.endofstream

    override val mime: String
        get() = packet.mime

    override val userFilename: String
    get() = packet.filename

    val autogenFilename: String
        get() {
            if (isEndOfStream) {
                return ""
            }

            val ext: String = getDefaultFileName(this) + "." + extension
            Log.e("debug", "getAutogenFilename: $ext")
            return ext
        }

    override val type: PacketType
        get() = PacketType.TYPE_BLOCKHEADER

    /**
     * Gets hash.
     *
     * @param seqnum the seqnum
     * @return the hash
     */
    fun getHash(seqnum: Int): ByteString {
        return packet.getNexthashes(seqnum)
    }

    /**
     * The type Builder.
     */
    data class Builder(
            private  var toDisk: Boolean = false,
            private var application: String = "",
            private var sessionid: Int = -1,
            private var mToFingerprint: ArrayList<UUID> = arrayListOf(),
            private var mFromFingerprint: ArrayList<UUID> = arrayListOf(),
            private var extensionVal: String = "",
            private var hashlist: List<ByteString> = ArrayList(),
            private var sig: ByteArray? = null,
            private var filename: String = "",
            private var mime: String = "",
            private var endofstream: Boolean = false,
            private var sendDate: Date = Date()
            ) {

        /**
         * Sets the fingerprint for the recipient.
         *
         * @param toFingerprint the to fingerprint
         * @return builder
         */
        fun setToFingerprint(toFingerprint: UUID?) = apply {
            if (toFingerprint != null) {
                mToFingerprint.add(toFingerprint)
            }
        }

        /**
         * Sets from fingerprint.
         *
         * @param fromFingerprint sets the fingerprint for the sender
         * @return builder
         */
        fun setFromFingerprint(fromFingerprint: UUID?) = apply {
            if (fromFingerprint != null) {
                mFromFingerprint.add(fromFingerprint)
            }
        }

        /**
         * Sets application.
         *
         * @param application bytes for UTF encoded scatterbrain application string
         * @return builder
         */
        fun setApplication(application: String) = apply {
            this.application = application
        }
        /**
         * Sets to disk.
         *
         * @param toDisk whether to write this file to disk or attempt to store it in the database
         * @return builder
         */
        fun setToDisk(toDisk: Boolean) = apply {
            this.toDisk = toDisk
        }

        /**
         * Sets session id.
         *
         * @param sessionID the session id (used for upgrading between protocols)
         * @return builder
         */
        fun setSessionID(sessionID: Int) = apply {
            sessionid = sessionID
        }

        /**
         * Sets hashes.
         *
         * @param hashes list of hashes of following blocksequence packets.
         * @return builder
         */
        fun setHashes(hashes: List<ByteString>) = apply {
            hashlist = hashes
        }

        /**
         * Sets the file extension
         * @param ext: string file extension
         * @return builder
         */
        fun setExtension(ext: String) = apply {
            this.extensionVal = sanitizeFilename(ext)
        }

        fun setSig(sig: ByteArray?) = apply {
            this.sig = sig
        }

        fun setMime(mime: String) = apply {
            this.mime = mime
        }

        fun setEndOfStream(value: Boolean) = apply {
            endofstream = value
        }

        fun setFilename(filename: String) = apply {
            this.filename = sanitizeFilename(filename)
        }

        fun setDate(date: Date) = apply {
            this.sendDate = date
        }

        /**
         * Build block header packet.
         *
         * @return the block header packet
         */
        fun build(): BlockHeaderPacket {
            val packet = BlockData.newBuilder()
                    .setApplication(application)
                    .addAllFromFingerprint(mFromFingerprint.map { u -> protoUUIDfromUUID(u) })
                    .addAllToFingerprint(mToFingerprint.map { u -> protoUUIDfromUUID(u) })
                    .setTodisk(toDisk)
                    .setExtension(extensionVal)
                    .addAllNexthashes(hashlist)
                    .setSessionid(sessionid)
                    .setSendDate(sendDate.time)
                    .setMime(mime)
                    .setTtl(0)
                    .setEndofstream(endofstream)
                    .setSig(ByteString.copyFrom(sig?: byteArrayOf(0)))
                    .build()

            return BlockHeaderPacket(packet)
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Builder

            if (toDisk != other.toDisk) return false
            if (!application.contentEquals(other.application)) return false
            if (sessionid != other.sessionid) return false

            if (extensionVal != other.extensionVal) return false
            if (hashlist != other.hashlist) return false
            if (sig != null) {
                if (other.sig == null) return false
                if (!sig.contentEquals(other.sig)) return false
            } else if (other.sig != null) return false
            if (filename != other.filename) return false
            if (mime != other.mime) return false
            if (endofstream != other.endofstream) return false

            return true
        }

        override fun hashCode(): Int {
            var result = toDisk.hashCode()
            result = 31 * result + sessionid
            result = 31 * result + extensionVal.hashCode()
            result = 31 * result + hashlist.hashCode()
            result = 31 * result + (sig?.contentHashCode() ?: 0)
            result = 31 * result + filename.hashCode()
            result = 31 * result + mime.hashCode()
            result = 31 * result + endofstream.hashCode()
            return result
        }

        /**
         * Instantiates a new Builder.
         */
        init {
            sessionid = -1
            mime = "application/octet-stream"
        }
    }

    companion object {
        /**
         * New builder builder.
         *
         * @return the builder
         */
        @kotlin.jvm.JvmStatic
        fun newBuilder(): Builder {
            return Builder()
        }

        class Parser : ScatterSerializable.Companion.Parser<BlockData, BlockHeaderPacket>(BlockData.parser())
        fun parser(): Parser {
            return Parser()
        }
    }
}